### ISSUE TITLE
class nmos::details::seed_generator is not a valid seed sequence using gcc 9.2.1

### ISSUE THREAD

**Issue by duncan-webb** [sony/nmos-cpp#124]
I'm raising this as it may prove to be a problem when building nmos-cpp containers using the upcoming Ubuntu 20.04 LTS base container image which installs gcc 9.2.1 by default. 

When compiling nmos/mdns.cpp the seed generator nmos::details::seed_generator used to seed the std::default_random_engine (a std::linear_congruential_engine) is failing a test in <random> (/usr/include/c++/9/bits/random.h):

```
/usr/include/c++/9/bits/random.h:293:9: note:   template argument deduction/substitution failed:
/usr/include/c++/9/bits/random.h: In substitution of ‘template<class _Sseq, class _Engine, class _Res, class _GenerateCheck> using __is_seed_seq = std::__and_<std::__not_<std::is_same<typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type, _Engine> >, std::is_unsigned<typename _Sseq::result_type>, std::__not_<std::is_convertible<_Sseq, _Res> > > [with _Sseq = nmos::details::seed_generator; _Engine = std::linear_congruential_engine<long unsigned int, 16807, 0, 2147483647>; _Res = long unsigned int; _GenerateCheck = void]’:
/usr/include/c++/9/bits/random.h:252:8:   required by substitution of ‘template<class _UIntType, _UIntType __a, _UIntType __c, _UIntType __m> template<class _Sseq> using _If_seed_seq = typename std::enable_if<std::__detail::__is_seed_seq<_Sseq, std::linear_congruential_engine<_UIntType, __a, __c, __m>, _UIntType>::value>::type [with _Sseq = nmos::details::seed_generator; _UIntType = long unsigned int; _UIntType __a = 16807; _UIntType __c = 0; _UIntType __m = 2147483647]’
/usr/include/c++/9/bits/random.h:291:32:   required from here
/usr/include/c++/9/bits/random.h:197:13: error: no type named ‘result_type’ in ‘class nmos::details::seed_generator’
  197 |       using __is_seed_seq = __and_<
      |             ^~~~~~~~~~~~~
```
The test that's failing in random.h that has been added sometime after 7.4:
```
    // Detect whether _Sseq is a valid seed sequence for
    // a random number engine _Engine with result type _Res.
    template<typename _Sseq, typename _Engine, typename _Res,
             typename _GenerateCheck = __seed_seq_generate_t<_Sseq>>
      using __is_seed_seq = __and_<
        __not_<is_same<__remove_cvref_t<_Sseq>, _Engine>>,
        is_unsigned<typename _Sseq::result_type>,
        __not_<is_convertible<_Sseq, _Res>>
      >;
```
I believe that this causes an earlier build error :
```
/nmos-cpp/Development/nmos/mdns.cpp: In lambda function:
/nmos-cpp/Development/nmos/mdns.cpp:577:101: error: no matching function for call to ‘std::linear_congruential_engine<long unsigned int, 16807, 0, 2147483647>::linear_congruential_engine(nmos::details::seed_generator&)’
  577 |                     std::shuffle(results->begin(), results->end(), std::default_random_engine(seeder));
      |                                                                                                     ^
```
It looks like the new random.h validation code is testing that the seed sequence class publicly defines the type result_type. Changing nmos::details::seed_generator to the following appears to keep cause it to pass the validation:
```
        class seed_generator
        {
        public:
            typedef std::uint32_t result_type;
            
            seed_generator() {}
            template <typename InputIterator> seed_generator(InputIterator first, InputIterator last) {}
            template <typename T> seed_generator(const std::initializer_list<T>&) {}

            template <typename RandomAccessorIterator>
            void generate(const RandomAccessorIterator first, const RandomAccessorIterator last)
            {
                std::uniform_int_distribution<result_type> uint32s;
                std::generate(first, last, [&] { return uint32s(device); });
            }

            size_t size() const { return 0; }
            template <typename OutputIterator> void param(OutputIterator) {}

        private:
            std::random_device device;
        };
```

**Comment by duncan-webb** [comment]
That seems to fit with the CPP reference for std::seed_seq (https://en.cppreference.com/w/cpp/numeric/random/seed_seq).

**Comment by duncan-webb** [comment]
It looks like gcc 9.2.1 is just stricter when checking for the SeedSequence named requirements now (https://en.cppreference.com/w/cpp/named_req/SeedSequence).

**Comment by garethsb-sony** [comment]
Thanks, @duncan-webb! Should be resolved by 7284ef8.
